### PR TITLE
update individual role secrets from infrastructure roles

### PR DESCRIPTION
**Pull request by alexeyklyukin**
When the password in the infrastructure role is updated, re-generate the
secret for that role.

Previously, the password for an infrastructure role was always fetched from
the secret, making any updates to such role a no-op after the corresponding
secret had been generated.

Addresses #149

### CODE REVIEW

File: pkg/cluster/sync.go [hunk 1]
```diff
@@ -331,8 +335,17 @@ func (c *Cluster) syncSecrets() error {
 				userMap = c.pgUsers
 			}
 			pwdUser := userMap[secretUsername]
-			pwdUser.Password = string(curSecret.Data["password"])
-			userMap[secretUsername] = pwdUser
+			// if this secret belongs to the infrastructure role and the password has changed - replace it in the secret
+			if pwdUser.Password != string(curSecret.Data["password"]) && pwdUser.Origin == spec.RoleOriginInfrastructure {
+				c.logger.Debugf("updating the secret %q from the infrastructure roles", secretSpec.Name)
+				if _, err := c.KubeClient.Secrets(secretSpec.Namespace).Update(secretSpec); err != nil {
+					return fmt.Errorf("could not update infrastructure role secret for role %q: %v", secretUsername, err)
```
Comment (zerg-junior) [inline pkg/cluster/sync.go]: Is it worth adding to this error message the infra secret name such as `postgresql-infrastructure-roles` ?

File: pkg/cluster/sync.go [hunk 2]
```diff
@@ -331,8 +335,17 @@ func (c *Cluster) syncSecrets() error {
...
+				}
+			} else {
+				// for non-infrastructure role - update the role with the password from the secret
```
Comment (zerg-junior) [inline pkg/cluster/sync.go]: 1. So the operator is supposed to forcibly overwrite the passwords using secrets if the passwords change for whatever reason ?
2. Are non-infra-role passwords supposed to change w/o changing the relevant secrets ? If not, does this code branch deserve a warning log statement ?

**Comment by zerg-junior** [comment]
As a general comment, I feel this change is worth (1) a test and (b) a note in the `readme`

File: pkg/cluster/sync.go (hunk 2 above)
Comment (Jan-M) [reply pkg/cluster/sync.go]: It is basically a feature for the infrastructure roles.

Comment (alexeyklyukin) [reply pkg/cluster/sync.go]: The only reason for changing the role's password could be changes in the infrastructure roles secret. The idea behind these PR is to enable anyone who has corresponding permissions to change the password for an infrastructure role throughout the cluster by simply changing it in the infrastructure roles secret, providing a single place for setting passwords (and other properties as well). Before that PR, your changes of the password in the infrastructure role had no effect unless you remove the corresponding secret generated for that role.

File: pkg/cluster/sync.go (hunk 1 above)
Comment (alexeyklyukin) [reply pkg/cluster/sync.go]: yes, good idea.

File: pkg/cluster/sync.go (hunk 2 above)
Comment (alexeyklyukin) [reply pkg/cluster/sync.go]: Answering question (2), the only user-definable non-infrastructure roles that have corresponding secrets are manifest roles, but we deliberately not allow to set passwords in the manifest. The only authoritative place to store those roles passwords are the corresponding secrets generated for them, so, contrary to the infrastructure roles, we don't have multiple places where we can set secrets and there is no need to sync them.

**Comment by Jan-M** [comment]
:+1:

**Comment by alexeyklyukin** [comment]
👍

**Comment by alexeyklyukin** [comment]
👍

**Comment by erthalion** [comment]
:+1:

**Pull request merged**